accepted/0000-workspaces-part2.md

```diff
@@ -0,0 +1,196 @@
+
+- `npm workspace foo test`
+- `npm ws foo test`
+- `npm :foo test`
+
+Leaving space to gather feedback from the community on whether to introduce one of this proposed commands or even introducing some novel interface.
+
+#### Install dependency example:
+
+Add `tap` as a **dependency** of a **workspace** named `foo`:
+
+```sh
+npm workspace foo install tap
+npm ws foo install tap
+npm :foo install tap
+```
+
+Note: **Globs** are not supported as a _subset/filter_ argument, the alternative is to use `npm workspace <alias> <cmd>` in which the `alias` can be defined beforehand in your **workspace** configuration as detailed in the next section.
+
+### 3. Grouping/Aliasing workspaces
+
+One proposed feature that was surfaced during our brainstorm sessions is the ability for users to create arbitrary groups of **workspaces** defined via **workspaces configuration**, e.g:
+
+Given a **npm workspace** setup with the following file structure:
+
+```
+/root
+├── core
+│   ├── foo
+│   ├── bar
+│   └── util
+└── plugins
+    ├── helpers
+    ├── lorem
+    └── ipsum
+```
+
+It's possible to define a **groups** property within your **workspaces configuration** that will allow defining named sets of **workspaces** that can be later on used to run commands, e.g:
+
+```
+{
+    "name": "workspace-example",
+    "version": "1.0.0",
+    "workspaces": {
+        "groups": {
+            "core": ["core/*"], // accepts fs location
+            "plugins": ["lorem", "ipsum"], // also accepts workspaces names
+            "common": ["util", "helpers"]
+        },
+        "packages": [
+            "core/*",
+            "plugins/*"
+        ]
+    }
+}
+```
+
+Following up with the previous configuration example, it would be possible to run tests across all **workspaces** from the `plugins` group, as following:
+
+```
+// one of these:
+$ npm workspace plugins test
+$ npm ws plugins test
+$ npm :plugins test
+
+> lorem@1.0.0 test /root/plugins/lorem
+> done
+
+> ipsum@1.0.0 test /root/plugins/ipsum
+> done
+```
+
+Further example, install a **peer dependency** across a group of packages named `core`:
+
+```
+$ npm workspace core install react@16 --save-peer
+$ npm ws core install react@16 --save-peer
+$ npm :core install react@16 --save-peer
+```
+
+## Prior Art
+
+#### Previously
+
+- [npm workspaces](https://github.com/npm/rfcs/blob/de8d71c0453f5cf443d3ef2f47e313f12dd6aaf9/accepted/0000-workspaces.md)
+
+#### Filtering a subset of workspaces
+
+- [lerna filter-options](https://www.npmjs.com/package/@lerna/filter-options)
+- [Yarn v1 workspace](https://classic.yarnpkg.com/en/docs/cli/workspace)
+- [Yarn v2 foreach include/exclude](https://yarnpkg.com/cli/workspaces/foreach)
+- [pnpm recursive --filter](https://pnpm.js.org/en/cli/recursive#filter-lt-package_selector)
```

Comment:
Looks like this moved to https://pnpm.js.org/en/filtering (I really like pnpm's approach, tbh)

Reply:
In the past, we used to also support passing the filters after `--`. So for instance, you could do `pnpm update -- foo bar` would update dependencies in `foo` and `bar`. I think that was pretty convinient but it caused confusion because some commands treat `--` differently.

I had some ideas [here](https://github.com/pnpm/pnpm/issues/2012) to make a unified syntax but in the end we decided to deprecate `--` and only allow passing the filters through one or more filter flags.

The filter flag can go before the command name, so you can run `pnpm --filter foo update`

Reply:
thanks for sharing that bit of context @zkochan 😊 

I see **pnpm** `--filter` also supports `<directory>` (which I suppose maybe also `globs` along with that?) what I proposed initially in this RFC is to avoid these in favor of a separate "aliases" OR "groups" definition that can also be defined in `package.json` for ease to share/reuse while also decoupling that categorization from the fs layout

Reply:
Directories must start with `./` or be inside `{}`

For instance, `--filter ./packages` or `--filter {packages}`. Also, they don't support globs. They select all packages inside the specified directories.

Reply:
also, you may combine selectors. Like you may filter by all packages that start with `babel-` and are inside some dir: `--filter babel-*{packages/subdir}`